pkg/devicetemplateconverter/profile_fetcher_test.go

```diff
@@ -0,0 +1,168 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicetemplateconverter_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/smartystreets/assertions"
+	"go.thethings.network/lorawan-stack/v3/pkg/cluster"
+	mockdr "go.thethings.network/lorawan-stack/v3/pkg/devicerepository/mock"
+	"go.thethings.network/lorawan-stack/v3/pkg/devicetemplateconverter"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/util/test"
+	"go.thethings.network/lorawan-stack/v3/pkg/util/test/assertions/should"
+	"google.golang.org/grpc"
+)
+
+func Test_profileFetcher_ValidateFields(t *testing.T) {
+	t.Parallel()
+	a := assertions.New(t)
+	tests := []struct {
+		name   string
+		device *ttnpb.EndDevice
+		want   bool
+	}{
+		{
+			name: "Fail/No_VendorID",
+			device: &ttnpb.EndDevice{
+				VersionIds: &ttnpb.EndDeviceVersionIdentifiers{
+					VendorProfileId: 1,
+				},
+			},
+			want: false,
+		},
+		{
+			name: "Fail/No_VendorProfileID",
+			device: &ttnpb.EndDevice{
+				VersionIds: &ttnpb.EndDeviceVersionIdentifiers{
+					VendorId: 1,
+				},
+			},
+			want: false,
+		},
+		{
+			name: "Valid",
+			device: &ttnpb.EndDevice{
+				VersionIds: &ttnpb.EndDeviceVersionIdentifiers{
+					VendorId:        1,
+					VendorProfileId: 1,
+				},
+			},
+			want: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(*testing.T) {
+			pf := &devicetemplateconverter.ProfileFetcher{}
+			a.So(pf.ValidateFields(tt.device), should.Equal, tt.want)
+		})
+	}
+}
+
+func Test_profileFetcher_FetchProfile(t *testing.T) {
+	t.Parallel()
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+	ctx, cancelCtx := context.WithCancel(ctx)
+	defer cancelCtx()
+
+	tests := []struct {
+		name         string
+		endDevice    *ttnpb.EndDevice
+		validateResp func(*assertions.Assertion, *ttnpb.EndDeviceTemplate)
+		validateErr  func(error) bool
+	}{
+		{
+			name: "fail/no_vendor_id",
```

Comment:
Just FYI - `t.Run` will automatically convert spaces to underlines, so you may just use spaces if you wish here.

Reply:
I know but I usually have the preference of explicitly typing the underscore. Happened automatically, will make it similar to the other tests naming.